Replace Conduction with Vibro

BPR02 opened the pull request:
- removes Conduction
- new shamir: Vibro, applied on boots
- when a player takes fall damage while sneaking, they will damage mobs in a 6 block radius
- damage dealt to each mob is determined by distance from the player and blocks between the player and mob
- sneaking on the ground will give incremental jump boost; it will wear off by the time they reach the ground again, so the attack can be used
    - 1-3 ticks: Jump Boost IV (4)
    - 4-6 ticks: Jump Boost VI (6)
    - 7-9 ticks: Jump Boost VIII (8)
    - 10+ ticks: Jump Boost X (10)
- when a player activates the shockwave, the 5x5 area directly below them will have a chance to apply desire lines effects
    - the center block is always affected
    - each block in the surrounding 3x3 has a 50% chance
    - each block in the 5x5 border has a 25% chance

BPR02 committed:
Replace Conduction with Vibro

- removes Conduction
- new shamir: Vibro, applied on boots
- when a player takes fall damage while sneaking, they will damage mobs in a 6 block radius
- damage dealt to each mob is determined by distance from the player and blocks between the player and mob
- sneaking on the ground will give 1 second of Jump Boost X
    - Vibro Jump deals 4 hearts of damage (if landing at the same altitude) with boots that don't have Feather Falling
    - Vibro Jump deals 2 hearts of damage with Feather Falling IV boots

BPR02 committed:
Increase Vibro Jump Boost

Jump Boost 10 was unreliable; sometimes the player's jump boost wouldn't run out and they wouldn't take damage. Jump Boost 11 guarantees (assuming little to no lag) that the jump boost wears out by the time they reach the ground

BPR02 committed:
Vibro - Incremental Jump and Desire Lines

- jump boost from Vibro is incremental, level increases based on how long you've been sneaking
    - 1-3 ticks: IV (4)
    - 4-5 ticks: VI (6)
    - 6-9 ticks: VII (8)
    - 10+ ticks: X (10)

- shockwave will have desire lines effects in a 5x5 area, if the module is installed

- removed version from drop to fix name

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/desire_lines.mcfunction:
```diff
@@ -0,0 +1,33 @@
+execute positioned ~1 ~ ~ if predicate gm4_vibro_shamir:desire_lines_chance_1 run function gm4_desire_lines:path
+execute positioned ~-1 ~ ~1 if predicate gm4_vibro_shamir:desire_lines_chance_1 run function gm4_desire_lines:path
+execute positioned ~ ~ ~1 if predicate gm4_vibro_shamir:desire_lines_chance_1 run function gm4_desire_lines:path
+execute positioned ~1 ~ ~1 if predicate gm4_vibro_shamir:desire_lines_chance_1 run function gm4_desire_lines:path
```
A better name for the predicate would be `desire_lines_chance_high`. This would also help reduce confusion between the two predicates in the future.

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/desire_lines.mcfunction:
```diff
@@ -0,0 +1,33 @@
+execute positioned ~-1 ~ ~2 if predicate gm4_vibro_shamir:desire_lines_chance_2 run function gm4_desire_lines:path
+execute positioned ~ ~ ~2 if predicate gm4_vibro_shamir:desire_lines_chance_2 run function gm4_desire_lines:path
+execute positioned ~1 ~ ~2 if predicate gm4_vibro_shamir:desire_lines_chance_2 run function gm4_desire_lines:path
+execute positioned ~2 ~ ~2 if predicate gm4_vibro_shamir:desire_lines_chance_2 run function gm4_desire_lines:path
```
Conversely, this predicate could be named `desire_lines_chance_low`.

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/jump.mcfunction:
```diff
@@ -0,0 +1,11 @@
+execute if score @s gm4_vibro_sneak matches 1 run playsound minecraft:block.stone.step block @s ~ ~ ~ 0.6 1.189207
+execute if score @s gm4_vibro_sneak matches 4 run playsound minecraft:block.stone.step block @s ~ ~ ~ 0.6 1.259921
+execute if score @s gm4_vibro_sneak matches 7 run playsound minecraft:block.stone.step block @s ~ ~ ~ 0.6 1.334840
+execute if score @s gm4_vibro_sneak matches 10 run playsound minecraft:block.stone.step block @s ~ ~ ~ 0.6 1.498307
```
I think it'd be ok to play these sounds to nearby players, not just `@s`. This would telegraph that the user is charging their jump, which might be useful to know in a PvP setting.

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/shock/user.mcfunction:
```diff
@@ -0,0 +1,22 @@
+scoreboard players operation $damage gm4_vibro_shock = @s gm4_vibro_shock
+
+# calculate damage dealt to mobs
+execute as @e[team=!invalid_team,type=!armor_stand,tag=!gm4_vibro_user,distance=..6] at @s run function gm4_vibro_shamir:shock/calculate
```
What's `team=!invalid_team`?

BPR02 replied:
That's used to check for a living mob, since the team selector only works for mobs and not other entities. This is used in liquid tanks, tho maybe we should prefix it `gm4_invalid_team` incase some outside module has a team named that?

BPR02 replied:
I've updated it to use a prefixed team name

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/shock/user.mcfunction:
```diff
@@ -0,0 +1,22 @@
+scoreboard players operation $damage gm4_vibro_shock = @s gm4_vibro_shock
+
+# calculate damage dealt to mobs
+execute as @e[team=!invalid_team,type=!armor_stand,tag=!gm4_vibro_user,distance=..6] at @s run function gm4_vibro_shamir:shock/calculate
```
I assume the armor stand exclusion is because they are technically a living mob but shouldn't actually be damaged. Would the use of an entity tag be worth it here? Should flying mobs be included in this?

BPR02 replied:
as written above, the `team=!invalid_team` is used to target living mobs. If we don't want flying mobs, then it'd be better to test for `nbt={OnGround:1b}`, so mobs not on the ground won't be affected. This would also make it so players can avoid the attack in a PVP setting

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/shock/user.mcfunction:
```diff
@@ -0,0 +1,22 @@
+particle minecraft:mycelium ~ ~ ~ 1 1 1 5 50
+particle dust_color_transition 0.239 0.251 0.247 1 1.000 0.996 0.961 ~ ~.2 ~ 2 0 2 0 200
+
+playsound minecraft:entity.generic.explode block @a[distance=..8] ~ ~ ~ 0.3 1.5
```
(Minor) This sound feels a bit too rumbly // lingers a bit too much in my opinion; maybe someone on the discord feels the same.

BPR02 replied:
Is it this sound or the totem of undying sound?

Bloo-dev reviewed (commented):
### In-Game Testing Results
Tested it in game and found one issue.
**Requirements**
- Boots of Ostara
- Desire Lines
- Vibro

**How to reproduce:**
1. Do a shockwave attack using vibro on some Boots of Ostara.
2. After the shockwave, switch the boots to some normal vibro boots. Do not move, just swap the boots.
3. Do a shockwave attack.
4. The desire lines effect will not kick in, even though you're using non-ostara boots. This only effects the first jump after the swap.

I also got this to trigger seemingly randomly, without the use of Boots of Ostara, but can't reproduce that reliably.

### Code Review
Looking at the code it seems pretty clean, too. Some of the Desire Lines related functions are missing header comments, though.
I also left some comments about some minor changes I'd make.

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/shock/calculate.mcfunction:
```diff
@@ -0,0 +1,18 @@
+
+# raycast to the vibro player
+summon marker ~ ~ ~ {CustomName:'"gm4_vibro_ray"',Tags:["gm4_vibro_ray"]}
+execute as @e[type=marker,tag=gm4_vibro_ray,limit=1] at @s facing entity @a[tag=gm4_vibro_user,limit=1] feet run tp @s ~ ~ ~ ~ ~
+scoreboard players reset $ray gm4_vibro_shock
+execute as @e[type=marker,tag=gm4_vibro_ray,limit=1] at @s run function gm4_vibro_shamir:shock/ray
+kill @e[type=marker,tag=gm4_vibro_ray]
```
You could cute the amount of `@e` calls by 66% if you moved these into a common function.

BPR02 replied:
This is how we've always done ray casting, but you're right; I'll update it

Bloo-dev commented on gm4_metallurgy/data/gm4_vibro_shamir/functions/shock/calculate.mcfunction:
```diff
@@ -0,0 +1,18 @@
+
+# if damage was more than .5 health (.25 hearts), damage the mob
+execute if score $shockwave gm4_vibro_shock matches 50.. unless entity @s[type=player] run function gm4_vibro_shamir:shock/apply_mob
+execute if score $shockwave gm4_vibro_shock matches 50.. if entity @s[type=player] run function gm4_vibro_shamir:shock/apply_player
```
This should probably ignore spectators and creative players.

BPR02 committed:
Bug Fixes, Optimizations

- fixes bug where Boots of Ostara could be switched out and still apply to normal vibro boots
- fixes bug where absorption damage doesn't get calculated
- fixes bug where creative and spectator players would get affected by the shockwave
- renamed desire lines chance predicates to be more clear
- moved ray casting to a helper function to reduce @e calls
- renamed the invalid team to use a gm4 prefix, in case another module uses that name
- plays charging sound to all nearby players instead of just the vibro user
